consensus/istanbul/backend/announce.go

```diff
@@ -589,130 +649,311 @@ func (sb *Backend) regossipAnnounce(msg *istanbul.Message, payload []byte) error
+	var validEntries []*vet.SignedAnnounceVersionEntry
+	validAddresses := make(map[common.Address]bool)
+	// Verify all entries are valid and remove duplicates
+	for _, signedAnnVersion := range signedAnnVersions {
+		err := signedAnnVersion.ValidateSignature()
+		if err != nil {
+			logger.Debug("Error validating signed announce version signature", "address", signedAnnVersion.Address, "err", err)
+			continue
+		}
+		if !validatorConnSet[signedAnnVersion.Address] {
+			logger.Debug("Found signed announce version from an address not in the validator conn set", "address", signedAnnVersion.Address)
+			continue
+		}
+		if _, ok := validAddresses[signedAnnVersion.Address]; ok {
+			logger.Debug("Found duplicate signed announce version in message", "address", signedAnnVersion.Address)
 			continue
 		}
+		validAddresses[signedAnnVersion.Address] = true
+		validEntries = append(validEntries, signedAnnVersion.Entry())
+	}
+	return sb.upsertSignedAnnounceVersionEntries(validEntries)
+}
 
-		sb.cachedAnnounceMsgsMu.RLock()
-		if cachedEntry, ok := sb.cachedAnnounceMsgs[announceVersion.ValAddress]; !ok || cachedEntry.MsgVersion < announceVersion.AnnounceMsgVersion {
-			announcesToRequest = append(announcesToRequest, announceVersion.ValAddress)
+func (sb *Backend) upsertSignedAnnounceVersionEntries(entries []*vet.SignedAnnounceVersionEntry) error {
+	logger := sb.logger.New("func", "upsertSignedAnnounceVersions")
+	newEntries, err := sb.signedAnnounceVersionTable.Upsert(entries)
+	if err != nil {
+		logger.Warn("Error in upserting entries", "err", err)
+	}
+
+	// If this node is a validator (checked later as a result of this call) and it receives a signed announce
+	// version from a remote validator that is newer than the remote validator's
+	// version in the val enode table, this node did not receive a direct announce
+	// and needs to announce its own enode to the remote validator.
+	sb.startGossipAnnounceTask()
+
+	// Only regossip entries that are new to this node and do not originate
+	// from an address that we have gossiped a signed announce version for
+	// within the last 5 minutes
+	var signedAnnVersionsToRegossip []*signedAnnounceVersion
+	sb.lastSignedAnnounceVersionsGossipedMu.Lock()
+	for _, entry := range newEntries {
+		lastGossipTime, ok := sb.lastSignedAnnounceVersionsGossiped[entry.Address]
+		if !ok || time.Since(lastGossipTime) >= 5*time.Minute {
+			signedAnnVersionsToRegossip = append(signedAnnVersionsToRegossip, &signedAnnounceVersion{
+				Address: entry.Address,
+				Version: entry.Version,
+				Signature: entry.Signature,
+			})
+			sb.lastSignedAnnounceVersionsGossiped[entry.Address] = time.Now()
 		}
-		sb.cachedAnnounceMsgsMu.RUnlock()
 	}
+	sb.lastSignedAnnounceVersionsGossipedMu.Unlock()
+	if len(signedAnnVersionsToRegossip) > 0 {
+		return sb.gossipSignedAnnounceVersionsMsg(signedAnnVersionsToRegossip)
+	}
+	return nil
+}
+
+func (sb *Backend) updateAnnounceVersion() {
```

Comment:
This function appears to be unused

Reply:
It's used in `addProxy` and `StartValidating`